### PR TITLE
add input validation

### PR DESCRIPTION
**Pull request by minz1027**
- Added input validation for start, stop and stream methods
- Re-wrote test cases to test the validation methods
- Resolved issue #5

### CODE REVIEW

File: test/index.test.js [hunk 1]
```diff
@@ -17,31 +17,86 @@ describe('index test', () => {
...
         const bar = new Foo({ foo: 'bar' });
 
         assert.instanceOf(bar, Executor);
-        bar.get(1, (err, data) => {
+        bar.stream({
+            buildId: 'a'
+        }, (err, data) => {
             assert.isNull(err);
-            assert.equal(data.id, 1);
+            assert.equal(data.buildId, 'b');
```
Comment (nkatzman) [inline test/index.test.js]: need to update this to `'a'`

File: index.js [hunk 2]
```diff
@@ -13,47 +15,123 @@ class Executor {
...
      * @param {String} config.buildId       Build id
      * @param {String} config.jobId         Job id
      * @param {String} config.pipelineId    Pipeline id
      * @param {String} config.container     Container
      * @param {String} config.scmUrl        Scm url
+     * @param {Function} callback           Function to call when done
+     */
+    start(config, callback) {
+        const schema = Joi.object().keys({
+            buildId: Joi.string().required(),
```
Comment (d2lam) [inline index.js]: Can we use data-schema for these?

File: test/index.test.js [hunk 3]
```diff
@@ -17,31 +17,86 @@ describe('index test', () => {
         assert.instanceOf(instance, Executor);
     });
 
-    it('has methods that need to be extended', () => {
-        assert.throws(instance.start, Error, 'not implemented');
-        assert.throws(instance.stop, Error, 'not implemented');
-        assert.throws(instance.stream, Error, 'not implemented');
+    it('start returns an error when not overridded', (done) => {
```
Comment (tkyi) [inline test/index.test.js]: overridden*

File: test/index.test.js [hunk 4]
```diff
@@ -17,31 +17,86 @@ describe('index test', () => {
...
+        };
+
+        instance.stop(config, (err) => {
+            assert.isOk(err, 'err is null');
+            assert.equal(err.message, 'not implemented');
+            done();
+        });
+    });
+
+    it('stream returns an error when not overridded', (done) => {
```
Comment (tkyi) [inline test/index.test.js]: overridden*

File: test/index.test.js (hunk 3 above)
Comment (d2lam) [inline test/index.test.js]: lol I blame @nkatzman

Comment (minz1027) [inline test/index.test.js]: Actually we discussed how to spell it, lol! working on it now.

**Comment by stjohnjohnson** [comment]
Looks like you are using the wrong email address for git commits.  Tisk tisk.

File: index.js [hunk 5]
```diff
@@ -1,4 +1,6 @@
 'use strict';
+const Joi = require('joi');
+/* eslint-disable no-underscore-dangle */
 /* eslint-disable no-console */
```
Comment (stjohnjohnson) [inline index.js]: This isn't needed anymore

File: index.js [hunk 6]
```diff
@@ -1,4 +1,6 @@
 'use strict';
+const Joi = require('joi');
```
Comment (stjohnjohnson) [inline index.js]: I would put this after the eslint comment

**Comment by minz1027** [comment]
@stjohnjohnson Got caught! Fixed!

File: index.js [hunk 7]
```diff
@@ -13,47 +14,123 @@ class Executor {
...
+    _start(config, callback) {
+        callback(new Error('not implemented'));
     }
 
     /**
-     * Stop the executor
+     * Validate the config for _stop method
      * @method stop
      * @param {Object} config               Configuration
      * @param {String} config.buildId       Build id
```
Comment (stjohnjohnson) [inline index.js]: Why does stop need more than a taskId?

Comment (d2lam) [inline index.js]: Same for `start`, we only need buildId and maybe container. k8s takes all the extra fields anyway (for labels) but it's not necessary.

Comment (stjohnjohnson) [inline index.js]: We should figure out exactly the fields required.  I don't think it should block this PR, but we should really solidify it this week.

**Comment by stjohnjohnson** [comment]
👍  
my_reaction_when_i_hit_refresh_again_and_that_upvote_disappears

**Comment by FenrirUnbound** [comment]
👍 

Seems reasonable to me

7lidvls - imgur

**Comment by d2lam** [comment]
👍  for now. We will probably need to move the validation into data-schema at some point.

**Pull request merged**